Partial results of running LOC8 to identify dead code.

zebmason opened the pull request:
You'll want to review these changes to ensure the code identified as dead is actually dead. Some of it looks like it is there for debug purposes.

The tool used to find the "dead" code is currently under development with a log at https://github.com/zebmason/LOC8. More work is needed on the tool to avoid false positives. If you don't find this useful let me know and I'll train the tool on other code.

dnfclas commented:
@__zebmason__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all Microsoft-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

dnfclas commented:
__@zebmason__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

jasonmalinowski commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerImageReference.cs:
```diff
@@ -65,27 +65,5 @@ public override object Id
             }
         }
-
-        private string GetDebuggerDisplay()
```
@zebmason This particular one is being used: the DebuggerDisplay attribute on the class will make this one matter.

jasonmalinowski commented:
Tagging @dotnet/roslyn-compiler, @dotnet/roslyn-ide and @dotnet/roslyn-analysis since some of these do look quite legitimate.

heejaechang commented:
@mavasani can you take a look?

mavasani commented on src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs:
```diff
@@ -137,13 +137,6 @@ private CompilationWithAnalyzers(Compilation compilation, ImmutableArray<Diagnos
             _executingCompilationOrNonConcurrentTreeTask = null;
         }
 
-        private void AddExceptionDiagnostic(Exception exception, DiagnosticAnalyzer analyzer, Diagnostic diagnostic)
```
So this seems to have uncovered a regression. This method should be hooked up in the constructor of CompilationWithAnalyzers and invoked when there is an exception from any analyzer. It seems this got regressed when we moved to AnalyzerDriver v2 in Dev14 Update3. Can you please retain this method and file a separate tracking issue to fix this and assign it to me?

tannergooding commented:
test windows_debug_vs-integration_prtest please

tannergooding commented:
test windows_release_vs-integration_prtest please

jcouv commented on src/Compilers/Core/CodeAnalysisTest/Resources.Designer.cs:
```diff
@@ -59,45 +59,5 @@ internal class Resources {
-        /// <summary>
-        ///   Looks up a localized resource of type System.Byte[].
-        /// </summary>
-        internal static byte[] defaultWin32Manifest {
```
I don't see the corresponding change to `resx` file (which generates this `cs` file).

heejaechang commented on src/Features/Core/Portable/Diagnostics/AnalyzerHelper.cs:
```diff
@@ -29,7 +29,6 @@ internal static class AnalyzerHelper
 
         // Shared with Compiler
         internal const string AnalyzerExceptionDiagnosticId = "AD0001";
-        internal const string AnalyzerDriverExceptionDiagnosticId = "AD0002";
```
this should be left since it is supposed to be in sync with the one defined in compiler side.

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder.cs:
```diff
@@ -776,66 +776,5 @@ internal BoundStatement WrapWithVariablesAndLocalFunctionsIfAny(CSharpSyntaxNode
             { WasCompilerGenerated = true };
         }
-
-        internal string Dump()
```
This is not dead code. We use it for debugging.

zebmason replied:
I was being a bit ruthless. A lot of the code which appears designed to be run in the debugger don't have #if DEBUG guards, perhaps they should?

CyrusNajmabadi replied:
No.  We still run things in the debugger when in release mode.  Now, maybe we should introduce an attribute to call out that this is intentional and is not dead code.  Then your tool could skip those guys.

zebmason replied:
Cyrus

Thanks for the feedback, the tool currently ignores build flavours and thus defines. Based on your comments I need to modify it so that it uses a tool specific build flavour in which all the relevant defines are specified.

Cheers

Zeb

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs:
```diff
@@ -379,14 +379,6 @@ internal static void FreeDeconstructionVariables(ArrayBuilder<DeconstructionVari
 
                 variables.Free();
             }
-            private string GetDebuggerDisplay()
```
Not dead.

jcouv commented on src/Compilers/CSharp/Portable/Parser/Directives.cs:
```diff
@@ -62,14 +62,6 @@ public bool IncrementallyEquivalent(Directive other)
         }
 
-        // Can't be private as it's called by DirectiveStack in its GetDebuggerDisplay()
-        internal string GetDebuggerDisplay()
```
None of the `GetDebbuggerDisplay` methods are dead code.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/NavigationBar/CSharpNavigationBarItemService.cs:
```diff
@@ -346,12 +346,6 @@ private static void ValidateSpanFromBounds(ITextSnapshot snapshot, int start, in
         }
 
-        [Conditional("DEBUG")]
-        private static void ValidateSpan(ITextSnapshot snapshot, int start, int length)
```
Are you testing with debug built?

zebmason replied:
Yes I built with debug, the analysis tool used needs a little work on it to select build flavours to ensure that when it uses Roslyn to create the semantic model that all the correct options are set.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/KeywordHighlighting/IfStatementHighlighterTests.cs:
```diff
@@ -131,28 +131,6 @@ public void Foo()
 }");
         }
 
-        private const string Code3 = @"
```
these worry me. t hey seem liek missed tests.

CyrusNajmabadi commented:
-18k lines is massive.  We're going to need this broken up.  perhaps it can be done so by grouping into folders.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/ImportChain.cs:
```diff
@@ -23,11 +23,6 @@ public ImportChain(Imports imports, ImportChain parentOpt)
-        {
-            return $"{Imports.GetDebuggerDisplay()} ^ {ParentOpt?.GetHashCode() ?? 0}";
-        }
-
```
This is not dead code.  It's referenced in the type attribute and used at debug time.  Same for every other GetDebuggerDisplay that is referenced.

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/CSharpResources.Designer.cs:
```diff
@@ -98,14026 +98,317 @@ internal class CSharpResources {
         }
         
         /// <summary>
-        ///   Looks up a localized string similar to Elements cannot be null..
```
This file has like 13k deletes.  Is that really supposed to be happening?  Also, we may want to just have a Pr that is the designer resources cleanup.

zebmason commented:
I've got the number of false positives produced by my tool down to less than 1700, still a lot to wade through and I can see some extra tests that I need to write. I've been a bit more brutal than before and deleted lots of code which appears to exist for use within the debugger - where it is marked as debug I've added #if DEBUG guards.

Obviously there are several possible reasons for the code that I've identified as being dead, examples being
1. It is there for debugging purposes
2. I didn't compile with the correct option to activate that code
3. It flags a regression
4. The code is partially completed and a subsequent commit will ensure that the dead code gets used
5. It wasn't tidied up after a commit

My tool uses Roslyn to parse the C# AST so I thought I might contribute something back by way of thanks. Hopefully the noise is minimal and there are some useful issues to address. I'll be quiet for a while now as I doubt that I've missed many members that my tool classifies as dead whilst I endeavour to productionise it for shipping.

zebmason commented:
@heejaechang the tool is currently written as a console application which is added as a pre-build step so that the warnings are printed to the output window to be clicked. It was my intention that once stabilised I turn it into a VS Extension and then sell it through VS Marketplace, however that is not currently a possibility. Apart from using Roslyn to parse C# (VB needs to be worked on) it uses libclang to parse C++ and its main purpose is for reviewing Lack of Cohesion problems - the dead code checking was only intended as a side function and I've been using it to verify that my C# parsing is correct.

Incidentally I've used it to refactor my own C# and C++ code - given that I was the only one to write that code it was embarrassing to find dead code and interesting to see where my design choices had reduced cohesion. On the latter point I've actually chopped up some classes so they make more sense and are more flexible.

CyrusNajmabadi commented:
Closing out.  It's not feasible to take this in in the current state (esp with all the conflicts now).  This is definitely an interesting thing to look at, but future PRs will have to be much smaller in scope (or broken into clear and easy commits to look at) to ensure that this is safe.
